toc/rfc/rfc-draft-bindable-secrets.md

```diff
@@ -0,0 +1,254 @@
+
+This command will ultimately create a `POST` request to `/v3/secrets`. Example `curl` invocation:
+
+```console
+curl "https://api.example.org/v3/secrets" \
+  -X POST \
+  -H "Authorization: bearer [token]" \
+  -H "Content-type: application/json" \
+  -d '{
+    "type": "json",
+    "name": "my-app-secrets",
+    "value": {
+      "application.yml" : "---\napplication:\n ...",
+      "secret.txt": "my-secret"
+    }
+    "metadata": {
+      "annotations": {},
+      "labels": {}
+    },
+    "relationships": {
+      "space": {
+        "data": {
+          "guid": "7304bc3c-7010-11ea-8840-48bf6bec2d78"
+        }
+      }
+    }
+  }'
+```
+
+### Reading a `secret` and Its Value
+Users that have permission to view environment variables for an app will be able to view `secret` values within a `space`. We will introduce a separate endpoint for retrieving these (similar to how app environment variables and service instance credentials are retrieved): `GET /v3/secrets/:guid/value`.
+
+```console
+curl "https://api.example.org/v3/secrets/[guid]/value" \
+  -X GET \
+  -H "Authorization: bearer [token]"
+  
+"value" : {
+  "some-json-key" : "some-json-value"
+}
+```
+
+To support this, Cloud Controller will need to have both read and write permissions for the `secret`'s corresponding credential in CredHub.
+
+### Creating a `secret_binding`
+To create `secret_bindings`, users will use a new CLI command: `cf bind-secret`. This will create a `POST` request to `/v3/secret_bindings`.
+
+```console
+cf bind-secret APP_NAME SECRET_NAME BINDING_NAME
+```
+
+### CLI Changes
+Work will need to be done to implement the following commands:
+- `cf create-secret`
+- `cf bind-secret`
+- `cf secrets`
+- `cf secret`
+- `cf delete-secret`
+- `cf unbind-secret`
+
+In addition, modifications may be made to existing CLI commands to surface `secrets` and `secret_bindings`. For example, adding bound `secret`s to `cf app` output.
+
+### Cloud Controller Changes
+New database tables will need to be created to store `secret`s and `secret_binding`s. CRUD APIs will need to be created for these resources. We will also need an API to support viewing a `secret`'s value.
+
+### BBS changes
+BBS was recently updated to support `ServiceBindingFiles` as part of RFC 0030. A `secret`'s value may result in one or more files being created in the `tmpfs` mounted directory defined by the `MountPath` parameter.
+
+```
+action := &models.RunAction{
+  Path: "/path/to/executable",
+  Args: []string{"some", "args to", "pass in"},
+  Dir: "/path/to/working/directory",
+  User: "username",
+  EnvironmentVariables: []*models.EnvironmentVariable{
+    {
+      Name: "ENVNAME",
+      Value: "ENVVALUE",
+    },
+  },
+ ServiceBindingFiles: []*models.Files{
+    {
+      Name: "/etc/cf-instance-binding",
+      Value: "VALUE",
+    },
+  },
+  Secrets: []*models.Secret{
+    {
+      Name: "SECRETNAME",
+      MountPath: "/etc/conf/",
```

Comment:
@winkingturtle-vmw @ameowlia 

Are there any paths we'd need to prevent here that would interfere with Diego or otherwise be bad? Allow app devs to drop arbitrary files _anywhere_ could be bad.

Reply:
Can we have the secret fail to mount + fail to start the app, if the mount path is already an existing directory on the container filesystem? 

These mounts would replace the directory structure they're mounted on top of, rather than be additive, right? If so, this should prevent most problems related to masking system directories/files, or accidentally masking app code or other required resources in the app container.

Reply:
❓ Could we make a dir where the secrets always go to? Where users could make nested dirs if they want. But this way we could limit exactly where they are putting things and not need to worry about it?